eth/db/atomic.py

```diff
@@ -31,69 +30,98 @@ def __init__(self, wrapped_db: BaseDB = None) -> None:
 
-        try:
-            with self._batch_lock:
-                yield
-        except Exception:
-            self.logger.exception(
-                "Unexpected error in atomic db write, dropped partial writes: %r",
-                self._diff(),
-            )
-            self._clear()
-            raise
-        else:
-            self._commit()
+
+class AtomicDBWriteBatch(BaseDB):
+    """
+    This is returned by a BaseAtomicDB during an atomic_batch, to provide a temporary view
+    of the database, before commit.
+    """
+    logger = logging.getLogger("eth.db.AtomicDBWriteBatch")
+
+    _write_target_db = None  # type: BaseDB
+    _track_diff = None  # type: DBDiffTracker
+
+    def __init__(self, _write_target_db: BaseDB) -> None:
+        self._write_target_db = _write_target_db
+        self._track_diff = DBDiffTracker()
 
     def __getitem__(self, key: bytes) -> bytes:
-        if not self._batch_lock.locked():
-            return self.wrapped_db[key]
+        if self._track_diff is None:
+            raise ValidationError("Cannot get data from a write batch, out of context")
 
         try:
             value = self._track_diff[key]
         except DiffMissingError as missing:
             if missing.is_deleted:
                 raise KeyError(key)
             else:
-                return self.wrapped_db[key]
+                return self._write_target_db[key]
         else:
             return value
 
     def __setitem__(self, key: bytes, value: bytes) -> None:
-        if self._batch_lock.locked():
-            self._track_diff[key] = value
-        else:
-            self.wrapped_db[key] = value
+        if self._track_diff is None:
+            raise ValidationError("Cannot set data from a write batch, out of context")
+
+        self._track_diff[key] = value
 
     def __delitem__(self, key: bytes) -> None:
+        if self._track_diff is None:
+            raise ValidationError("Cannot delete data from a write batch, out of context")
+
         if key not in self:
             raise KeyError(key)
-        if self._batch_lock.locked():
-            del self._track_diff[key]
-        else:
-            del self.wrapped_db[key]
+        del self._track_diff[key]
 
     def _diff(self) -> DBDiff:
         return self._track_diff.diff()
 
-    def _clear(self):
-        self._track_diff = DBDiffTracker()
-
     def _commit(self) -> None:
-        self._diff().apply_to(self.wrapped_db, apply_deletes=True)
-        self._clear()
+        self._diff().apply_to(self._write_target_db, apply_deletes=True)
 
     def _exists(self, key: bytes) -> bool:
+        if self._track_diff is None:
+            raise ValidationError("Cannot test data existance from a write batch, out of context")
+
         try:
             self[key]
         except KeyError:
             return False
         else:
             return True
+
+    @classmethod
+    @contextmanager
+    def commit_unexceptional(cls, write_target_db):
```

Comment:
This name is odd, but am I correct that it's an internal API.  If so, then I don't really care that much other than maybe prefixing with an `_` to make that clear.